File: sdk/python/clearcote/geoip.py
"""geoip: resolve the egress IP's geo (timezone + lat/lon + language) so the browser matches the
proxy actually in use — the way Camoufox does it. Primary source is daijro's offline
"geoip-all-in-one" MaxMind DB (more accurate than a single online API; merges IP2Location +
GeoLite2 + DB-IP, timezone computed from coordinates). Flow: discover the exit IP via a small
IP-echo *through the proxy*, then look that IP up in the cached .mmdb. Falls back to ip-api.com
(direct geo through the proxy) if the DB can't be fetched/opened.

The .mmdb (GPL-3.0 data) is downloaded + cached on first use (~52 MB zip -> ~120 MB), NOT bundled.
The exit-IP and ip-api lookups go THROUGH the proxy (HTTP or SOCKS5, see ``_net``); we never fall
back to the local IP under a proxy, which would give the wrong region. Needs the `maxminddb`
package (a dependency); if it's missing, falls back to ip-api.com.

Budget: CLEARCOTE_GEOIP_TIMEOUT_SECONDS (default 20) bounds the whole resolution. A lookup that
runs out of budget fails rather than hanging a launch.
"""

import json
import math
import os
import shutil
import sys
import tempfile
import threading
import time
import urllib.request
import zipfile

from ._net import proxied_request, to_proxy_spec

try:
    import maxminddb  # type: ignore
    _HAVE_MAXMINDDB = True
except ImportError:
    _HAVE_MAXMINDDB = False

MMDB_URL = "https://github.com/daijro/geoip-all-in-one/releases/latest/download/geoip-aio-all.mmdb.zip"
MMDB_MAX_AGE_DAYS = 30
IPECHO_URLS = ("http://icanhazip.com", "http://api.ipify.org", "http://ip-api.com/line/?fields=query")
IPAPI_URL = "http://ip-api.com/json/?fields=status,message,countryCode,timezone,lat,lon,query"
# dotted record paths for the geoip-all-in-one schema (GeoLite2-City shaped)
_PATHS = {"iso_code": "country.iso_code", "longitude": "location.longitude",
          "latitude": "location.latitude", "timezone": "location.time_zone"}


def _log(quiet, msg):
    if not quiet:
        sys.stderr.write(f"[clearcote] {msg}\n")
        sys.stderr.flush()


def _geo_cache_root():
    """Same location the Node SDK uses, so a downloaded DB is shared."""
    env = os.environ.get("CLEARCOTE_CACHE")
    if env:
        return os.path.join(env, "geoip")
    if sys.platform == "win32":
        base = os.environ.get("LOCALAPPDATA") or os.path.expanduser(r"~\AppData\Local")
        return os.path.join(base, "clearcote", "geoip")
    if sys.platform == "darwin":
        return os.path.join(os.path.expanduser("~/Library/Caches"), "clearcote", "geoip")
    base = os.environ.get("XDG_CACHE_HOME") or os.path.expanduser("~/.cache")
    return os.path.join(base, "clearcote", "geoip")


def geo_cache_root():
    """Where the geoip database is cached (CLEARCOTE_CACHE/geoip when set)."""
    return _geo_cache_root()


def geoip_timeout_seconds(env=None):
    """Whole-resolution budget in seconds: CLEARCOTE_GEOIP_TIMEOUT_SECONDS (> 0), default 20."""
    env = os.environ if env is None else env
    raw = str(env.get("CLEARCOTE_GEOIP_TIMEOUT_SECONDS") or "").strip()
    try:
        n = float(raw)
    except ValueError:
        return 20.0
    if not math.isfinite(n) or n <= 0:
        return 20.0
    return n


class GeoipError(RuntimeError):
    """Raised by launch() when ``geoip=True`` was requested and the region could not be resolved.

    Failing closed is the point: continuing would launch with the host's clock and a default
    language -- UTC + en-US on most servers -- which is exactly the mismatch geoip exists to prevent.
    Pass an explicit ``timezone`` AND ``accept_language`` to launch anyway when the lookup is
    unavailable.
    """

    code = "GEOIP_UNRESOLVED"


def _find_in(data, dotted):
    for part in dotted.split("."):
        if not isinstance(data, dict):
            return None
        data = data.get(part)
        if data is None:
            return None
    return data


def _get_text(url, spec, timeout):
    res = proxied_request(url, timeout=max(0.001, timeout), proxy=spec)
    if not res.ok:
        raise RuntimeError(f"HTTP {res.status}")
    return res.text().strip()


def _looks_like_ip(s):
    if s.count(".") == 3 and all(p.isdigit() for p in s.split(".")):
        return True
    return ":" in s and all(c in "0123456789abcdefABCDEF:" for c in s)


def _exit_ip(spec, deadline, quiet):
    for url in IPECHO_URLS:
        left = deadline - time.monotonic()
        if left <= 0:
            break
        try:
            text = _get_text(url, spec, min(left, 8.0))
            ip = text.split()[0].strip() if text else ""
            if _looks_like_ip(ip):
                return ip
        except Exception:  # noqa: BLE001
            continue
    return None


def _ensure_mmdb(quiet):
    if not _HAVE_MAXMINDDB:
        return None
    d = _geo_cache_root()
    mmdb = os.path.join(d, "geoip-aio-all.mmdb")
    if os.path.exists(mmdb):
        age_days = (time.time() - os.path.getmtime(mmdb)) / 86400.0
        if age_days < MMDB_MAX_AGE_DAYS:
            return mmdb
    try:
        os.makedirs(d, exist_ok=True)
        _log(quiet, "geoip: downloading the geoip-all-in-one database (~52 MB, first run only)")
        with tempfile.NamedTemporaryFile(suffix=".zip", delete=False) as tmp:
            zpath = tmp.name
        part = mmdb + ".part-%d" % os.getpid()
        try:
            with urllib.request.urlopen(  # noqa: S310
                urllib.request.Request(MMDB_URL, headers={"User-Agent": "clearcote-sdk"}), timeout=120
            ) as resp, open(zpath, "wb") as out:
                shutil.copyfileobj(resp, out)
            with zipfile.ZipFile(zpath) as z:
                member = next((n for n in z.namelist() if n.lower().endswith(".mmdb")), None)
                if not member:
                    raise RuntimeError("no .mmdb in archive")
                with z.open(member) as src, open(part, "wb") as dst:
                    shutil.copyfileobj(src, dst)
            os.replace(part, mmdb)  # atomic: a timed-out waiter never opens a half-written file
        finally:
            for f in (zpath, part):
                try:
                    os.remove(f)
                except OSError:
                    pass
        _log(quiet, "geoip: database ready")
        return mmdb
    except Exception as e:  # noqa: BLE001
        _log(quiet, f"geoip: database fetch failed ({type(e).__name__}: {e}) — falling back to ip-api")
        return None


_MMDB_LOCK = threading.Lock()
_MMDB_THREAD = None


def _ensure_mmdb_within(quiet, left):
    """_ensure_mmdb bounded by ``left`` seconds. The first-run database download (~52 MB) may
    outlast the budget: it keeps going in a background thread and caches for the next launch,
    while this launch falls back to ip-api instead of waiting."""
    global _MMDB_THREAD
    result = {}
    with _MMDB_LOCK:
        if _MMDB_THREAD is None or not _MMDB_THREAD.is_alive():
            def run():
                result["path"] = _ensure_mmdb(quiet)
            _MMDB_THREAD = threading.Thread(target=run, name="clearcote-geoip-db", daemon=True)
            _MMDB_THREAD.start()
        thread = _MMDB_THREAD  # possibly a download already in flight from an earlier launch
    thread.join(max(0.0, left))
    if thread.is_alive():
        return None
    if "path" in result:
        return result["path"]
    mmdb = os.path.join(_geo_cache_root(), "geoip-aio-all.mmdb")
    return mmdb if os.path.exists(mmdb) else None


def _mmdb_lookup(ip, deadline, quiet):
    left = deadline - time.monotonic()
    if left <= 0:
        return None
    mmdb = _ensure_mmdb_within(quiet, left)
    if not mmdb:
        return None
    try:
        with maxminddb.open_database(mmdb) as reader:
            rec = reader.get(ip)
        if not rec:
            return None
        country = _find_in(rec, _PATHS["iso_code"])
        lat = _find_in(rec, _PATHS["latitude"])
        lon = _find_in(rec, _PATHS["longitude"])
        tz = _find_in(rec, _PATHS["timezone"])
        if not tz and lat is None:
            return None
        return {
            "ip": ip,
            "country": (str(country).upper() if country else None),
            "timezone": (str(tz) if tz else None),
            "accept_language": accept_language_for_country(country),
            "location": (f"{lat},{lon}" if lat is not None and lon is not None else None),
        }
    except Exception as e:  # noqa: BLE001
        _log(quiet, f"geoip: mmdb read failed ({type(e).__name__}: {e})")
        return None


def _ip_api_fallback(spec, deadline):
    left = deadline - time.monotonic()
    if left <= 0:
        return None
    try:
        j = json.loads(_get_text(IPAPI_URL, spec, min(left, 8.0)))
        if j.get("status") != "success":
            return None
        lat, lon = j.get("lat"), j.get("lon")
        return {
            "ip": j.get("query"),
            "country": j.get("countryCode"),
            "timezone": j.get("timezone"),
            "accept_language": accept_language_for_country(j.get("countryCode")),
            "location": (f"{lat},{lon}" if lat is not None and lon is not None else None),
        }
    except Exception:  # noqa: BLE001
        return None


def resolve_geo_detailed(proxy=None, quiet=False, timeout=None):
    """Resolve geo for the egress (through ``proxy`` if given -- HTTP or SOCKS5, a Playwright dict
    or a URL string -- else direct), reporting why it failed.

    Returns ``(geo, reason, elapsed_ms)``: ``geo`` is the dict (or None), ``reason`` a human-readable
    failure reason when there is no geo/timezone (else None). Never raises. Bounded by ``timeout``
    seconds (default :func:`geoip_timeout_seconds`)."""
    started = time.monotonic()
    budget = float(timeout) if timeout is not None else geoip_timeout_seconds()
    deadline = started + budget

    def elapsed():
        return int(round((time.monotonic() - started) * 1000))

    try:
        spec = to_proxy_spec(proxy) if proxy else None
    except Exception as e:  # noqa: BLE001
        return None, f"invalid proxy ({e})", elapsed()
    ip = _exit_ip(spec, deadline, quiet)
    geo = _mmdb_lookup(ip, deadline, quiet) if ip else None
    if not geo:
        geo = _ip_api_fallback(spec, deadline)
    if geo and geo.get("timezone"):
        _log(quiet, f"geoip: {geo['ip']} -> {geo['country']} "
                    f"tz={geo['timezone']} lang={geo['accept_language']}")
        return geo, None, elapsed()
    if time.monotonic() >= deadline:
        secs = "%g" % round(budget, 1)
        reason = f"timed out after {secs}s (CLEARCOTE_GEOIP_TIMEOUT_SECONDS)"
    elif not ip:
        reason = "could not determine the exit IP" + (" through the proxy" if spec else "")
    elif geo:
        reason = f"no timezone for exit IP {ip}"
    else:
        reason = f"no geo data for exit IP {ip}"
    return geo, reason, elapsed()


def resolve_geo(proxy=None, quiet=False, timeout=None):
    """Resolve geo for the egress (through ``proxy`` if given, else direct). Never raises — returns
    None on failure. geoip-all-in-one offline DB first, ip-api.com fallback. Returns a dict
    {ip, country, timezone, accept_language, location} or None. ``timeout`` bounds the whole
    resolution in seconds (default CLEARCOTE_GEOIP_TIMEOUT_SECONDS, else 20)."""
    return resolve_geo_detailed(proxy, quiet=quiet, timeout=timeout)[0]


# country (ISO-3166 alpha-2) -> Accept-Language. Plain comma list (NO ;q= weights). The geoip DB
# has no language data, so this maps the resolved country. Falls back to en-US,en.
COUNTRY_LANG = {
    "US": "en-US,en", "GB": "en-GB,en", "CA": "en-CA,en,fr-CA", "AU": "en-AU,en", "NZ": "en-NZ,en",
    "IE": "en-IE,en", "IN": "en-IN,en,hi", "ZA": "en-ZA,en", "SG": "en-SG,en",
    "DE": "de-DE,de,en", "AT": "de-AT,de,en", "CH": "de-CH,de,fr,en",
    "FR": "fr-FR,fr,en", "BE": "nl-BE,nl,fr,en", "NL": "nl-NL,nl,en",
    "ES": "es-ES,es,en", "MX": "es-MX,es,en", "AR": "es-AR,es,en", "CL": "es-CL,es,en",
    "CO": "es-CO,es,en", "PT": "pt-PT,pt,en", "BR": "pt-BR,pt,en",
    "IT": "it-IT,it,en", "PL": "pl-PL,pl,en", "RU": "ru-RU,ru,en", "UA": "uk-UA,uk,ru,en",
    "SE": "sv-SE,sv,en", "NO": "nb-NO,no,en", "DK": "da-DK,da,en", "FI": "fi-FI,fi,en",
    "CZ": "cs-CZ,cs,en", "RO": "ro-RO,ro,en", "HU": "hu-HU,hu,en", "GR": "el-GR,el,en",
    "TR": "tr-TR,tr,en", "IL": "he-IL,he,en", "SA": "ar-SA,ar,en", "AE": "ar-AE,ar,en",
    "EG": "ar-EG,ar,en", "JP": "ja-JP,ja,en", "KR": "ko-KR,ko,en",
    "CN": "zh-CN,zh,en", "HK": "zh-HK,zh,en", "TW": "zh-TW,zh,en",
    "TH": "th-TH,th,en", "VN": "vi-VN,vi,en", "ID": "id-ID,id,en",
    "MY": "ms-MY,ms,en", "PH": "en-PH,en,fil",
}


def accept_language_for_country(cc):
    if not cc:
        return "en-US,en"
    return COUNTRY_LANG.get(str(cc).upper(), "en-US,en")

File: sdk/node/src/geoip.ts
// geoip: resolve the egress IP's geo (timezone + lat/lon + language) so the browser matches the
// proxy actually in use — the way Camoufox does it. Primary source is daijro's offline
// "geoip-all-in-one" MaxMind DB (more accurate than a single online API; merges IP2Location +
// GeoLite2 + DB-IP, timezone computed from coordinates). Flow: discover the exit IP via a small
// IP-echo *through the proxy*, then look that IP up in the cached .mmdb. Falls back to ip-api.com
// (direct geo through the proxy) if the DB can't be fetched/opened.
//
// The .mmdb (GPL-3.0 data) is downloaded + cached on first use (≈52 MB zip → ≈120 MB), NOT bundled.
// The exit-IP and ip-api lookups go THROUGH the proxy (HTTP or SOCKS5, see ./net.ts); we never fall
// back to the local IP under a proxy, which would give the wrong region.
//
// Budget: CLEARCOTE_GEOIP_TIMEOUT_SECONDS (default 20) bounds the whole resolution. A lookup that
// runs out of budget fails rather than hanging a launch.

import { createWriteStream, existsSync, mkdirSync, readdirSync, rmSync, statSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import extract from "extract-zip";
import maxmind, { type Reader } from "maxmind";
import { proxiedRequest, toProxySpec, type ProxySpec } from "./net.js";

const MMDB_URL = "https://github.com/daijro/geoip-all-in-one/releases/latest/download/geoip-aio-all.mmdb.zip";
const MMDB_MAX_AGE_DAYS = 30;
const IPECHO_URLS = ["http://icanhazip.com", "http://api.ipify.org", "http://ip-api.com/line/?fields=query"];
const IPAPI_URL = "http://ip-api.com/json/?fields=status,message,countryCode,timezone,lat,lon,query";

export interface Geo {
  ip?: string;
  country?: string;
  timezone?: string;
  acceptLanguage?: string;
  location?: string; // "lat,lon"
}

function log(quiet: boolean | undefined, m: string): void {
  if (!quiet) process.stderr.write(`[clearcote] ${m}\n`);
}

/** Where the geoip database is cached (CLEARCOTE_CACHE/geoip when set). */
export function geoCacheRoot(): string {
  if (process.env.CLEARCOTE_CACHE) return path.join(process.env.CLEARCOTE_CACHE, "geoip");
  if (process.platform === "win32")
    return path.join(process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local"), "clearcote", "geoip");
  if (process.platform === "darwin") return path.join(os.homedir(), "Library", "Caches", "clearcote", "geoip");
  return path.join(process.env.XDG_CACHE_HOME || path.join(os.homedir(), ".cache"), "clearcote", "geoip");
}

/** Whole-resolution budget in ms: CLEARCOTE_GEOIP_TIMEOUT_SECONDS (seconds, > 0), default 20s. */
export function geoipTimeoutMs(env: Record<string, string | undefined> = process.env): number {
  const raw = (env.CLEARCOTE_GEOIP_TIMEOUT_SECONDS ?? "").trim();
  const n = Number(raw);
  return raw && Number.isFinite(n) && n > 0 ? Math.round(n * 1000) : 20_000;
}

async function getText(url: string, proxy: ProxySpec | null, timeoutMs: number): Promise<string> {
  const res = await proxiedRequest(url, { proxy, timeoutMs: Math.max(1, timeoutMs) });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return (await res.text()).trim();
}

const IPV4 = /^(?:\d{1,3}\.){3}\d{1,3}$/;
const IPV6 = /^[0-9a-f:]+$/i;
function looksLikeIp(s: string): boolean {
  return IPV4.test(s) || (s.includes(":") && IPV6.test(s));
}

/** Discover the egress IP via an IP-echo through the proxy (or direct), within `deadline`. */
async function exitIp(proxy: ProxySpec | null, deadline: number, quiet?: boolean): Promise<string | null> {
  for (const url of IPECHO_URLS) {
    const left = deadline - Date.now();
    if (left <= 0) break;
    try {
      const ip = (await getText(url, proxy, Math.min(left, 8000))).split(/\s+/)[0];
      if (looksLikeIp(ip)) return ip;
    } catch {
      /* try next */
    }
  }
  log(quiet, "geoip: could not determine the exit IP");
  return null;
}

let _mmdbInflight: Promise<string | null> | null = null;
async function ensureMmdb(quiet?: boolean): Promise<string | null> {
  const dir = geoCacheRoot();
  const file = path.join(dir, "geoip-aio-all.mmdb");
  if (existsSync(file)) {
    const ageDays = (Date.now() - statSync(file).mtimeMs) / 86_400_000;
    if (ageDays < MMDB_MAX_AGE_DAYS) return file;
  }
  if (_mmdbInflight) return _mmdbInflight;
  _mmdbInflight = (async () => {
    try {
      mkdirSync(dir, { recursive: true });
      const zip = path.join(dir, "geoip-aio-all.mmdb.zip");
      log(quiet, "geoip: downloading the geoip-all-in-one database (~52 MB, first run only)");
      const res = await fetch(MMDB_URL, { redirect: "follow" });
      if (!res.ok || !res.body) throw new Error(`HTTP ${res.status}`);
      await pipeline(Readable.fromWeb(res.body as any), createWriteStream(zip));
      const tmp = path.join(dir, ".extract");
      rmSync(tmp, { recursive: true, force: true });
      await extract(zip, { dir: tmp });
      const found = findMmdb(tmp);
      if (!found) throw new Error("no .mmdb in archive");
      rmSync(file, { force: true });
      // move (rename across same dir)
      const fs = await import("node:fs");
      fs.renameSync(found, file);
      rmSync(zip, { force: true });
      rmSync(tmp, { recursive: true, force: true });
      log(quiet, "geoip: database ready");
      return file;
    } catch (e) {
      log(quiet, `geoip: database fetch failed (${(e as Error).message}) — falling back to ip-api`);
      return null;
    } finally {
      _mmdbInflight = null;
    }
  })();
  return _mmdbInflight;
}

function findMmdb(dir: string): string | null {
  const stack = [dir];
  while (stack.length) {
    const cur = stack.pop() as string;
    for (const e of readdirSync(cur, { withFileTypes: true })) {
      const full = path.join(cur, e.name);
      if (e.isDirectory()) stack.push(full);
      else if (e.name.toLowerCase().endsWith(".mmdb")) return full;
    }
  }
  return null;
}

let _reader: Reader<any> | null = null;
async function mmdbLookup(ip: string, deadline: number, quiet?: boolean): Promise<Geo | null> {
  // The first-run database download (~52 MB) may outlast the budget. It keeps going in the
  // background and caches for the next launch; this launch falls back to ip-api instead of waiting.
  const left = deadline - Date.now();
  if (left <= 0) return null;
  let timer: NodeJS.Timeout | undefined;
  const file = await Promise.race([
    ensureMmdb(quiet),
    new Promise<null>((resolve) => { timer = setTimeout(() => resolve(null), left); }),
  ]);
  if (timer) clearTimeout(timer);
  if (!file) return null;
  try {
    if (!_reader) _reader = await maxmind.open(file);
    const rec: any = _reader.get(ip);
    if (!rec) return null;
    const country: string | undefined = rec?.country?.iso_code;
    const lat = rec?.location?.latitude;
    const lon = rec?.location?.longitude;
    const tz: string | undefined = rec?.location?.time_zone;
    if (!tz && lat == null) return null;
    return {
      ip,
      country,
      timezone: tz,
      acceptLanguage: acceptLanguageForCountry(country),
      location: lat != null && lon != null ? `${lat},${lon}` : undefined,
    };
  } catch (e) {
    log(quiet, `geoip: mmdb read failed (${(e as Error).message})`);
    return null;
  }
}

// Fallback: ip-api.com returns geo directly (through the proxy), no DB needed.
async function ipApiFallback(proxy: ProxySpec | null, deadline: number): Promise<Geo | null> {
  const left = deadline - Date.now();
  if (left <= 0) return null;
  try {
    const txt = await getText(IPAPI_URL, proxy, Math.min(left, 8000));
    const j = JSON.parse(txt);
    if (j?.status !== "success") return null;
    return {
      ip: j.query,
      country: j.countryCode,
      timezone: j.timezone,
      acceptLanguage: acceptLanguageForCountry(j.countryCode),
      location: j.lat != null && j.lon != null ? `${j.lat},${j.lon}` : undefined,
    };
  } catch {
    return null;
  }
}

/** Outcome of a geo resolution: the geo, or why there is none. */
export interface GeoResult {
  geo: Geo | null;
  /** Human-readable failure reason when `geo` is null or has no timezone. */
  reason?: string;
  /** How long the resolution took, in ms. */
  elapsedMs: number;
}

/**
 * Resolve geo for the egress (through `proxy` if given — HTTP or SOCKS5 — else direct), reporting
 * why it failed. Never throws. Bounded by `timeoutMs` (default {@link geoipTimeoutMs}).
 */
export async function resolveGeoDetailed(
  proxy?: string | { server?: string; username?: string; password?: string },
  opts: { quiet?: boolean; timeoutMs?: number } = {},
): Promise<GeoResult> {
  const started = Date.now();
  const budget = opts.timeoutMs ?? geoipTimeoutMs();
  const deadline = started + budget;
  let spec: ProxySpec | null;
  try {
    spec = toProxySpec(proxy ?? null);
  } catch (e) {
    return { geo: null, reason: `invalid proxy (${(e as Error).message})`, elapsedMs: Date.now() - started };
  }
  const ip = await exitIp(spec, deadline, opts.quiet);
  let geo: Geo | null = ip ? await mmdbLookup(ip, deadline, opts.quiet) : null;
  if (!geo) geo = await ipApiFallback(spec, deadline);
  const elapsedMs = Date.now() - started;
  if (geo && geo.timezone) {
    log(opts.quiet, `geoip: ${geo.ip} -> ${geo.country} tz=${geo.timezone} lang=${geo.acceptLanguage}`);
    return { geo, elapsedMs };
  }
  const reason = Date.now() >= deadline
    ? `timed out after ${Math.round(budget / 1000)}s (CLEARCOTE_GEOIP_TIMEOUT_SECONDS)`
    : !ip
      ? `could not determine the exit IP${spec ? " through the proxy" : ""}`
      : geo
        ? `no timezone for exit IP ${ip}`
        : `no geo data for exit IP ${ip}`;
  return { geo, reason, elapsedMs };
}

/**
 * Resolve geo for the egress (through `proxy` if given, else direct). Never throws — returns null
 * on failure. Uses the geoip-all-in-one offline DB first, ip-api.com as a fallback.
 */
export async function resolveGeo(
  proxy?: string | { server?: string; username?: string; password?: string },
  opts: { quiet?: boolean; timeoutMs?: number } = {}
): Promise<Geo | null> {
  return (await resolveGeoDetailed(proxy, opts)).geo;
}

// country (ISO-3166 alpha-2) -> Accept-Language. Plain comma list (NO ;q= weights — Chromium's
// --accept-lang adds those). The geoip DB has no language data, so this maps the resolved country.
const COUNTRY_LANG: Record<string, string> = {
  US: "en-US,en", GB: "en-GB,en", CA: "en-CA,en,fr-CA", AU: "en-AU,en", NZ: "en-NZ,en",
  IE: "en-IE,en", IN: "en-IN,en,hi", ZA: "en-ZA,en", SG: "en-SG,en",
  DE: "de-DE,de,en", AT: "de-AT,de,en", CH: "de-CH,de,fr,en",
  FR: "fr-FR,fr,en", BE: "nl-BE,nl,fr,en", NL: "nl-NL,nl,en",
  ES: "es-ES,es,en", MX: "es-MX,es,en", AR: "es-AR,es,en", CL: "es-CL,es,en",
  CO: "es-CO,es,en", PT: "pt-PT,pt,en", BR: "pt-BR,pt,en",
  IT: "it-IT,it,en", PL: "pl-PL,pl,en", RU: "ru-RU,ru,en", UA: "uk-UA,uk,ru,en",
  SE: "sv-SE,sv,en", NO: "nb-NO,no,en", DK: "da-DK,da,en", FI: "fi-FI,fi,en",
  CZ: "cs-CZ,cs,en", RO: "ro-RO,ro,en", HU: "hu-HU,hu,en", GR: "el-GR,el,en",
  TR: "tr-TR,tr,en", IL: "he-IL,he,en", SA: "ar-SA,ar,en", AE: "ar-AE,ar,en",
  EG: "ar-EG,ar,en", JP: "ja-JP,ja,en", KR: "ko-KR,ko,en",
  CN: "zh-CN,zh,en", HK: "zh-HK,zh,en", TW: "zh-TW,zh,en",
  TH: "th-TH,th,en", VN: "vi-VN,vi,en", ID: "id-ID,id,en",
  MY: "ms-MY,ms,en", PH: "en-PH,en,fil",
};

export function acceptLanguageForCountry(cc?: string): string {
  if (!cc) return "en-US,en";
  return COUNTRY_LANG[cc.toUpperCase()] || "en-US,en";
}

/**
 * Thrown by launch() when `geoip: true` was requested and the region could not be resolved.
 *
 * Failing closed is the point: continuing would launch with the host's clock and a default
 * language — UTC + en-US on most servers — which is exactly the mismatch geoip exists to prevent.
 * Pass an explicit `timezone` AND `acceptLanguage` to launch anyway when the lookup is unavailable.
 */
export class GeoipError extends Error {
  code = "GEOIP_UNRESOLVED";
  constructor(message: string) {
    super(message);
    this.name = "GeoipError";
  }
}
